### ISSUE TITLE
ValueError: Inconsistent code

### ISSUE THREAD

**Issue by GoogleCodeExporter** [tallforasmurf/byteplay#5]
```
What steps will reproduce the problem?

$ python byteplay.py testcase.py
Traceback (most recent call last):
  File "byteplay.py", line 924, in <module>
    main()
  File "byteplay.py", line 921, in main
    recompile_all(sys.argv[1])
  File "byteplay.py", line 900, in recompile_all
    recompile(filename)
  File "byteplay.py", line 878, in recompile
    codeobject2 = cod.to_code()
  File "byteplay.py", line 649, in to_code
    co_stacksize = self._compute_stacksize()
  File "byteplay.py", line 642, in _compute_stacksize
    open_positions.extend(get_next_stacks(pos, curstack))
  File "byteplay.py", line 492, in get_next_stacks
    raise ValueError, "Inconsistent code"
ValueError: Inconsistent code

What version of the product are you using? On what operating system?
byteplay 0.2, Python 2.7.6, Debian Linux
```

Original issue reported on code.google.com by `jwilk@jwilk.net` on 17 Jan 2014 at 9:04

Attachments:
* [testcase.py](https://storage.googleapis.com/google-code-attachments/byteplay/issue-5/comment-0/testcase.py)

**Comment by tallforasmurf** [comment]
def ttt():
      while 1:
       try:
            try:
                continue
            finally:
                pass
       finally:
           pass

This fails in byteplay3 also.
HAH! In order to make it executable I replaced `continue` with `break` and it no longer fails!

That points us to the following, 

    elif op == CONTINUE_LOOP:
                # CONTINUE_LOOP jumps to the beginning of a loop which should
                # already have been discovered, but we verify anyway.
                # It pops a block.
                yield label_pos[arg], curstack[:-1]

I had noticed that blithe comment "It pops a block" before and wondered. Because I do not think it is true. `continue` by itself doesn't pop a loop, it merely jumps to the top of the loop where the condition may or may not result in a jump to the loop terminus. In this case, not.

**Comment by tallforasmurf** [comment]
Using a more readable test,

    def ttt():
        for i in range(2):
            try:
                try:
                    continue
                finally:
                    pass
            finally:
                pass

This produces the error. Again, replacing `continue` with `break` eliminates the error. Why is BREAK_LOOP given a `pass` in the code, and CONTINUE_LOOP treated as ending a block? Here is the implementation:

        TARGET(BREAK_LOOP) {
            why = WHY_BREAK;
            goto fast_block_end;
        }

        TARGET(CONTINUE_LOOP) {
            retval = PyLong_FromLong(oparg);
            if (retval == NULL)
                goto error;
            why = WHY_CONTINUE;
            goto fast_block_end;
        }

Thus both have the ability to end a block but when we get to fast_block_end,

            if (b->b_type == SETUP_LOOP && why == WHY_CONTINUE) {
                why = WHY_NOT;
                JUMPTO(PyLong_AS_LONG(retval));
                Py_DECREF(retval);
                break;
            }
            /* Now we have to pop the block. */
            f->f_iblock--;

In other words, if we come from WHY_CONTINUE we end up doing a jump to the oparg, which is the opcode address for the top of the loop. But if we come in with WHY_BREAK, "we have to pop the block".

I am going to try reversing the treatment of CONTINUE_LOOP and BREAK_LOOP opcodes.

**Comment by tallforasmurf** [comment]
Simpler failing case:

    def t2():
      while True:
        try:
          continue
        finally:
          pass

It is necessary to have the continue inside a try (`finally:` is not required, `except:` will do as well)

**Comment by tallforasmurf** [comment]
Discovered that in many cases, Python 3 converts `continue` into JUMP_ABSOLUTE. When it is not, it has the destination arg of the loop top FOR_ITER or (in the failing case) SETUP_FINALLY. It should not send the stack size calculator back there. So hopefully the fix, which is to just `pass` CONTINUE_LOOP, will not have downstream problems.
